src/ZkFold/Symbolic/Compiler/Arithmetizable.hs

```diff
@@ -15,102 +16,98 @@ import           Numeric.Natural                                     (Natural)
 import           Prelude                                             hiding (Num (..), drop, length, product, splitAt,
                                                                       sum, take, (!!), (^))
 
-import           ZkFold.Base.Algebra.Basic.Class
 import           ZkFold.Base.Algebra.Basic.Number
-import           ZkFold.Base.Data.Vector                             hiding (concat)
-import           ZkFold.Prelude                                      (drop, length, splitAt, take)
-import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal (Arithmetic, ArithmeticCircuit)
+import qualified ZkFold.Base.Data.Vector                             as V
+import           ZkFold.Base.Data.Vector                             (Vector (..))
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal (Arithmetic, ArithmeticCircuit (..), Circuit,
+                                                                      concatCircuits, joinCircuits)
 
 -- | A class for Symbolic data types.
 -- Type `a` is the finite field of the arithmetic circuit.
 -- Type `x` represents the data type.
-class Arithmetic a => SymbolicData a x where
-    -- | Returns the circuits that make up `x`.
-    pieces :: x -> [ArithmeticCircuit a]
+class (Arithmetic a, KnownNat n) => SymbolicData a n x | x a -> n where
+    -- | Returns the circuit that makes up `x`.
+    pieces :: x -> ArithmeticCircuit n a
 
-    -- | Restores `x` from the circuits' outputs.
-    restore :: [ArithmeticCircuit a] -> x
+    -- | Restores `x` from the circuit's outputs.
+    restore :: Circuit a -> Vector n Natural -> x
 
     -- | Returns the number of finite field elements needed to describe `x`.
     typeSize :: Natural
+    typeSize = value @n
```

Comment:
Any other implementation would not make sense, so we better make this a separate function instead of a class member